### PR TITLE
Adding base classes for WT3 functionality

### PR DESCRIPTION
**Pull request by andrewlee94**
## Summary/Motivation:
This PR adds base classes for WT3 type property packages and zero-order unit models along with associated tests.

## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/core/zero_order_properties.py
Comment (adam-a-a) [inline watertap/core/zero_order_properties.py:278]: Do we need the `p` here?

Comment (andrewlee94) [reply watertap/core/zero_order_properties.py:278]: Yes, we do. It is a part of the framework API, so when this method is called it will always be passed `p` (even though we don't actually need it in this case).

File: watertap/core/zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/zero_order_unit.py]: In the NF and RO unit models, we've used `recovery_vol_phase` for volumetric recovery (the typical form) and `recovery_mass_phase_comp` for mass recovery of each component. In this case, `recovery_vol_phase` is the equivalent. Perhaps we want to keep this as is at the moment for the WT3 migration to prevent potential issues with renaming across files that make use of `water_recovery`, but just wanted to make note of that so that we are all aware.

File: watertap/core/zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/zero_order_unit.py]: I'm not sure that `deltaP_outlet` makes sense to have in all cases. For example, with RO, `deltaP_outlet` would be the pressure at the permeate outlet (typically atmospheric or slightly higher and on one side of the membrane if considering a flat sheet element) minus the feed pressure at the inlet. @kurbansitterley would probably have some better insight on when/where it makes sense to have both `deltaP_outlet` and `deltaP_waste` for WT3 models, but from my perspective, it seems like we would only be concerned with `deltaP_waste`. If that is true, we could eliminate `deltaP_outlet` and eventually rename `deltaP_waste` as `deltaP`.

File: watertap/core/zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/zero_order_unit.py]:
```suggestion
        # Temperature equality
```

File: watertap/core/zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/zero_order_unit.py:202]: For the case of NF/RO, I don't think we would need this.

File: watertap/core/zero_order_unit.py
Comment (andrewlee94) [reply watertap/core/zero_order_unit.py]: I am not sure that is a good idea - there could be cases where we need both (or even cases where we want `deltaP_outlet` and not `deltaP_waste`). However, there are ways we can make these terms optional, and let the user or the derived classes decide which terms to include.

Two ideas I have has so far:

1. Have `has_pressure_change` config arguments for each outlet, and rely on the user to set them. This has some minor issues for cases where a term is not optional however, as the config argument will still exist (but only have one valid value).
2. Have a private attribute on the unit model which the derived class sets before calling `super().build()`. This hides the option from the user and lets the unit model developer make the decision, which is probably a better option as the unit model would also then know whether it needs to look up a value for `deltaP` from the database. The downside is that the unit model developer needs to be aware of this and set the private attribute correctly.

File: watertap/core/zero_order_unit.py
Comment (andrewlee94) [reply watertap/core/zero_order_unit.py:202]: This ties into the comment above on whether and how `deltaP` should be an optional term in these types of units.

File: watertap/core/zero_order_unit.py
Comment (andrewlee94) [reply watertap/core/zero_order_unit.py]: For now, I went with the WT3 naming convention (for the most part), but we should think about standardizing naming.

File: watertap/core/zero_order_unit.py
Comment (kurbansitterley) [reply watertap/core/zero_order_unit.py]: Yes, in most cases the pressure on the permeate side will be atmospheric, but I can think of two scenarios where you might want a `deltaP_outlet` variable:

- In a multi-pass system where the permeate from the first RO unit goes to another RO unit. This is done for example when the Boron concentration in the permeate requires further reduction to meet a regulation. Something like this could require a booster pump in between the first and second passes, so you might want to know the pressure differential coming out of the permeate just to be consistent in the momentum balance.
- If you want to use "permeate throttling" as a way to direct/control flow between stages. This is where you would intentionally throttle pressure on the permeate side, thus reducing flux for that module and sending more flow to the next stage. I am honestly not very familiar with this configuration/operation and only became acquainted with it because it was used at Kay Bailey (one of the baseline studies; we did not include permeate throttling in our model) so I looked it up. Attached figure and more information on permeate throttling from [here](https://membranes.com/docs/trc/flowcon.pdf).

Screen Shot 2021-11-16 at 12 51 08 PM

My vote would be to at least have the ability to have both, but regardless to keep the `_waste` in the variable name (i.e. `deltaP_waste` rather than `deltaP`) just in the interest of being explicit. I can't really comment on the implementation for watertap but I hope this is helpful.

File: watertap/core/zero_order_unit.py
Comment (andrewlee94) [reply watertap/core/zero_order_unit.py]: I changes these to `recovery_vol` and `removal_mass_solute` to better fit the naming convention. Seeing as these models only support 1 phase, I did not include the phase index.

File: watertap/core/zero_order_unit.py
Comment (adam-a-a) [reply watertap/core/zero_order_unit.py]: After reading your comments, I think it is best to keep the two options, and it looks like @andrewlee94 implemented his second option. With regard to RO, @kurbansitterley, I think these examples would be resolved by specifying the permeate outlet pressure explicitly rather than the pressure change between feed inlet and permeate outlet. Nevertheless, keeping both deltaP variables seems like the way to go.

File: docs/technical_reference/core/zero_order_unit.rst
Comment (adam-a-a) [inline docs/technical_reference/core/zero_order_unit.rst]:
```suggestion
The zero-order unit model base class is intended to form the basis for a library of generic single inlet-two outlet (SITO) unit models using recovery and removal fractions to partition the incoming flow between the two outlets. This base class is not intended to be used by itself and requires the user to define a derived class before it can be used in a flowsheet.
```

File: docs/technical_reference/core/zero_order_unit.rst
Comment (adam-a-a) [inline docs/technical_reference/core/zero_order_unit.rst]:
```suggestion
The SITO base class writes the following constraints which relate the inlet state to those in the outlet and waste streams. As mentioned previously, these constraints are formulated such that they will be linear if `flow_vol`, `conc_mass_comp`, `pressure` and `temperature` are the state variables used by the property package and the recovery and removal fractions are fixed.
```

File: watertap/core/tests/test_zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/tests/test_zero_order_unit.py]:
```suggestion
                           "package. Zero-order models only support property "
```

File: watertap/core/tests/test_zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/tests/test_zero_order_unit.py]:
```suggestion
                       match="fs.unit derived class has not been implemented "
```

File: watertap/core/tests/test_zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/tests/test_zero_order_unit.py]:
```suggestion
                       match="fs.unit derived class has not been implemented "
```

File: watertap/core/tests/test_zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/tests/test_zero_order_unit.py:292]: Can use `assert_optimal_termination(results)` method from `pyomo.environ` instead

File: watertap/core/tests/test_zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/tests/test_zero_order_unit.py:630]: Same as above (although this works and no dire need to change if you don't want to)

File: watertap/core/zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/zero_order_unit.py]:
```suggestion
    cannot be instantiated by itself. When creating derived classes,
```

File: watertap/core/zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/zero_order_unit.py]:
```suggestion
    determine whether the deltaP terms for the outlet and waste streams are
```

File: watertap/core/zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/zero_order_unit.py]:
```suggestion
                f"{self.name} derived class has not been implemented "
```

File: watertap/core/zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/zero_order_unit.py]:
```suggestion
                f"{self.name} derived class has not been implemented "
```

File: watertap/core/zero_order_unit.py
Comment (adam-a-a) [inline watertap/core/zero_order_unit.py]:
```suggestion
                f"Zero-order models only support property packages with a "
```

**Review by adam-a-a — COMMENTED** [review]
@andrewlee94 Just a few typos that I fixed and a couple quick, optional changes suggested

**Comment by adam-a-a** [comment]
If the penultimate commit (c8cd71f) fails, then it is simply due to fixing typos in a test and missing the typo in either the property or unit model base class

**Review by adam-a-a — APPROVED** [review]
LGTM - up to @andrewlee94 if you care to throw in the `assert_optimal_termination` method in tests, but it shouldn't hold up the PR.

**Review by kurbansitterley — APPROVED** [review]
I am not familiar with the testing files so I was mostly looking at the `zero_order_unit` ... I would wait for another reviewer before a merge.

**Pull request merged**
